src/window.rs

```diff
@@ -60,16 +58,46 @@ impl Window {
             fb
         };
 
-        if !self.crtc_set {
-            self.crtc_set = true;
+        if self.frame_count == 0 {
+            // NOTE(mbernat): It's possible to avoid initial mode setting
+            // since we're keeping the previous mode: we could just call page_flip directly.
             self.drm_display.set_mode_with_framebuffer(Some(fb));
         } else {
             self.drm_display.page_flip(fb);
         }
+
+        // NOTE(mbernat): This buffer object returns to the surface's queue upon destruction.
+        // If we were to release it here, it would be again available from
+        // `Surface::lock_front_buffer()` next and the app would be effectively single-buffered.
+        // So, we keep it around for one frame, which is fine for double buffering.
+        self.previous_bo.replace(buffer_object);
+
+        self.frame_count += 1;
     }
 
     pub fn restore_original_display(&self) {
-        self.drm_display.set_mode_with_framebuffer(self.drm_display.crtc.framebuffer());
+        let handle = self.drm_display.crtc.framebuffer().unwrap();
+        if self.drm_display.gbm_device.get_framebuffer(handle).is_ok() {
+            self.drm_display.set_mode_with_framebuffer(self.drm_display.crtc.framebuffer());
```

Comment:
```suggestion
            self.drm_display.set_mode_with_framebuffer(Some(handle));
```

Reply:
Alternatively, we could change `set_mode_with_framebuffer()` to take `FramebufferHandle` instead of `Option<FramebufferHandle>`, since the function name implies we'll always be working with a framebuffer.

And in the two cases where we currently call it, we always have `Some(handle)`.